### ISSUE TITLE
btrfs balance on mixed capacity drives

### ISSUE THREAD

**Issue by CHerSun** [kdave/btrfs-progs#802]
Hi. Sorry if this is a duplicate, but I wasn't able to find issue on this.

I'm using RAID1 on 5 drives, all of the drives were 3 TB. Recently I've had to switch a failed drive to a 4 TB one. And only then I noted, that btrfs balances by free space on the drive and not on percentage used (used/total) of the drive. Between 3 and 4 TB drives that won't cause too large skew, but googling I've noted cases where ppl use configurations like 1+1+3+3 drives, and there this causes a huge skew, forcing to basically 2 disks until ~2 TB is written. I'm considering to upgrade to ~18 TB drives one by one - then I will get this skew.

So the question is if it's possible to switch to percentage-based balancing. If not - could that be added?

**Comment by Forza-tng** [comment]
You can reduce the usable size of the 4th disk to 3TiB by using:

```
btrfs filesystem resize <id>:<size> <mount-point>
```
You can list device id using `btrfs fi us -T <mnt>`

**Comment by CHerSun** [comment]
> You can reduce the usable size of the 4th disk to 3TiB by using:

Thank you for idea, but well, that's not why one would increase drives capacity and use btrfs, quite the opposite.

**Comment by Forza-tng** [comment]
> > You can reduce the usable size of the 4th disk to 3TiB by using:
> 
> Thank you for idea, but well, that's not why one would increase drives capacity and use btrfs, quite the opposite.

Well, perhaps i misunderstood. Btrfs only uses 2 mirrors for any allocation with RAID1. So more disks will not make things faster or more resilient. With RAID10, Btrfs would use stripes across devices, potentially making sequential reads/writes faster.

**Comment by CHerSun** [comment]
> Well, perhaps i misunderstood. Btrfs only uses 2 mirrors for any allocation with RAID1. So more disks will not make things faster or more resilient. With RAID10, Btrfs would use stripes across devices, potentially making sequential reads/writes faster.

Yep, 2 mirrors for **single** block, but next block can use 2 mirrors on different disks, etc. So more disks = more speed, especially for larger transfers (considering 1GiB allocations). Placing all of allocations on larger disks gives a huge skew to both possible performance and drives wear. That's why most hardware storages, that do allow mixing capacities normally do distrubution of allocations based on % - to utilize all of them more evenly.

**Comment by Forza-tng** [comment]
It would be a good option, had it existed. For now, the easiest way to do this is to use slack, and then reduce as needed.
